columnar/src/columnar/merge/mod.rs

```diff
@@ -348,56 +358,157 @@ fn is_empty_after_merge(
-    // For merging we are interested in the same column type category since they can be merged.
-    let mut columns_grouped: HashMap<(String, ColumnTypeCategory), GroupedColumns> = HashMap::new();
-
-    for &(ref column_name, column_type) in required_columns {
-        columns_grouped
-            .entry((column_name.clone(), column_type.into()))
-            .or_insert_with(|| {
-                GroupedColumns::for_category(column_type.into(), columnar_readers.len())
-            })
-            .require_type(column_type)?;
+type MergeIter<'a> =
+    Box<dyn Iterator<Item = io::Result<(Rc<str>, ColumnType, GroupedColumns)>> + 'a>;
+
+#[derive(Debug, Clone)]
+struct MergeColumn {
+    column_name: Rc<str>,
+    reader_ord: usize,
+    column: DynamicColumnHandle,
+}
+impl MergeColumn {
+    fn new(column_name: Rc<str>, reader_ord: usize, column: DynamicColumnHandle) -> Self {
+        MergeColumn {
+            column_name,
+            reader_ord,
+            column,
+        }
     }
+}
 
-    for (columnar_id, columnar_reader) in columnar_readers.iter().enumerate() {
-        let column_name_and_handle = columnar_reader.list_columns()?;
-        // We skip columns that end up with 0 documents.
-        // That way, we make sure they don't end up influencing the merge type or
-        // creating empty columns.
-
-        for (column_name, handle) in column_name_and_handle {
-            let column_category: ColumnTypeCategory = handle.column_type().into();
-            let column = handle.open()?;
-            if is_empty_after_merge(merge_row_order, &column, columnar_id) {
-                continue;
+/// Iterates over the columns of the columnar readers, grouped by column name.
+/// Key functionality is that `open` of the Columns is done lazy per group.
+fn group_columns_for_merge_iter<'a>(
+    columnar_readers: &'a [&'a ColumnarReader],
+    required_columns: &'a [(String, ColumnType)],
+    merge_row_order: &'a MergeRowOrder,
+) -> io::Result<impl Iterator<Item = io::Result<(Rc<str>, ColumnType, GroupedColumns)>> + 'a> {
+    let column_iters: Vec<_> = columnar_readers
+        .iter()
+        .enumerate()
+        .map(|(reader_ord, reader)| {
+            Ok(reader
+                .iter_columns()?
+                .map(move |el| MergeColumn::new(Rc::from(el.0), reader_ord, el.1)))
+        })
+        .collect::<io::Result<_>>()?;
+    let required_columns_map: HashMap<String, ColumnType> = required_columns
+        .iter()
+        .map(|(col_name, typ)| (col_name.to_string(), *typ))
+        .collect::<HashMap<String, _>>();
+    let mut required_columns_list: Vec<String> = required_columns
+        .iter()
+        .map(|(col_name, _)| col_name.to_string())
+        .collect();
+    required_columns_list.sort();
+
+    // Kmerge and group on column_name.
+    let group_iter = GroupByIteratorExtended::group_by(
+        column_iters
+            .into_iter()
+            .kmerge_by(|a, b| a.column_name < b.column_name),
```

Comment:
nitpick... I prefer `lhs, rhs` or `left`,`right` for that comparisons.

```suggestion
            .kmerge_by(|a, b| a.column_name < b.column_name),
```